### PR TITLE
[candi] fix migrate cluster from one edition to another

### PR DESCRIPTION
**Pull request by mvasl**
## Description

Moved rendering of kubernetes-api-proxy manifest to a later point in time during bashible execution. This way, by the time manifest is rendered, we already know how to properly login into registry. `bashible-apiserver` is now stopped before applying new registry by scaling down it's `Deployment` to 0 replicas.

## Why do we need it, and what problem does it solve?

This fixes a bug where a cluster cannot migrate between community and enterprise editions properly, falling apart in the process due to losing kube-apiserver connectivity after new images for edition are pulled and migration process is started.

## Checklist
- [ ] The code is covered by unit tests.
- [x] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: candi
type: fix
summary: fix migration of cluster from one edition to another.
impact_level: default
```

### CODE REVIEW

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @m-vasilenko. I've started the workflow for label 'e2e/run/openstack'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @m-vasilenko. I've started the workflow for label 'e2e/run/static'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @m-vasilenko. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @m-vasilenko. I've started the workflow for label 'e2e/run/static'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @m-vasilenko. I've started the workflow for label 'e2e/run/static'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @m-vasilenko. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @m-vasilenko. I've started the workflow for label 'e2e/run/openstack'...

File: candi/bashible/common-steps/node-group/051_pull_kubernetes_api_proxy.sh.tpl
Comment (RomanenkoDenys) [inline candi/bashible/common-steps/node-group/051_pull_kubernetes_api_proxy.sh.tpl:1]: please rename file to 051_pull_and_configure_kubernetes_api_proxy.sh.tpl

**Review by RomanenkoDenys — COMMENTED** [review]
all ok, but please rename one file

File: candi/bashible/common-steps/node-group/051_pull_and_configure_kubernetes_api_proxy.sh.tpl [hunk 1]
```diff
@@ -0,0 +1,54 @@
+# Copyright 2021 Flant JSC
```
Comment (RomanenkoDenys) [inline candi/bashible/common-steps/node-group/051_pull_and_configure_kubernetes_api_proxy.sh.tpl:1]:
```suggestion
# Copyright 2023 Flant JSC
```

**Review by RomanenkoDenys — APPROVED** [review]
LGTM

File: deckhouse-controller/pkg/helpers/change_registry/change_registry.go [hunk 2]
```diff
@@ -113,10 +113,17 @@ func ChangeRegistry(newRegistry, username, password, caFile, newDeckhouseImageTa
...
 		deploymentYaml, _ := yaml.Marshal(deckhouseDeploy)
 		logEntry.Println("------------------------------")
+		logEntry.Println("Deployment bashible-apiserver will be scaled to 0 replicas")
+		logEntry.Println("------------------------------")
 		logEntry.Printf("New Secret will be applied:\n%s\n", secretYaml)
 		logEntry.Println("------------------------------")
 		logEntry.Printf("New Deployment will be applied:\n%s\n", deploymentYaml)
 	} else {
+		logEntry.Println("Scaling-down bashible-apiserver...")
+		if err := scaleBashibleAPIServerDeployment(ctx, kubeCl, 0); err != nil {
```
Comment (yalosev) [inline deckhouse-controller/pkg/helpers/change_registry/change_registry.go:123]: Looks weird. I really don't like this solution.
What for do we scale the deployment? We have a lock during the rebuilt phase, maybe we just should add some sleep there. Or fix the basic problem with reloading?
Scaling down and up is really a crutch. Also I don't see back-scaling, I think it should be done by Deckhouse

Comment (yalosev) [reply deckhouse-controller/pkg/helpers/change_registry/change_registry.go:123]: We have discussed it with @RomanenkoDenys.
Let's create an issue, to replace this part of code with `Deckhouse Pod readiness` subscription in the bashibe-apiserver. But we could do it later, it's not urgent.

**Review by yalosev — APPROVED** [review]
Please, create a follow-up issue about pod's subscription before merging

**Comment by deckhouse-BOaTswain** [comment]
Cherry pick PR [5746](https://github.com/deckhouse/deckhouse/pull/5746) to the branch [release-1.50](https://github.com/deckhouse/deckhouse/tree/release-1.50) successful!

**Pull request merged**